src/test/java/com/fasterxml/jackson/databind/deser/TestKeyDeserializerOverwritten.java

```diff
@@ -0,0 +1,65 @@
+package com.fasterxml.jackson.databind.deser;
+
+import java.io.IOException;
+import java.util.Map;
+
+import org.junit.jupiter.api.Test;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.KeyDeserializer;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.module.SimpleModule;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+// [databind#4444]
+public class TestKeyDeserializerOverwritten {
+    @JsonDeserialize(keyUsing = ForClass.class)
+    static class MyKey {
+        private final String value;
+
+        MyKey(String value) {
+            this.value = value;
+        }
+    }
+
+    static class ForClass extends KeyDeserializer {
+        @Override
+        public Object deserializeKey(String key, DeserializationContext ctxt) throws IOException {
+            return new MyKey(key + "-class");
+        }
+    }
+
+    static class ForMapper extends KeyDeserializer {
+        @Override
+        public Object deserializeKey(String key, DeserializationContext ctxt) throws IOException {
+            return new MyKey(key + "-mapper");
+        }
+    }
+
+    // It is not declared as new TypeReference<> because it causes a compile error in Java 8.
+    TypeReference<Map<MyKey, String>> typeRef = new TypeReference<Map<MyKey, String>>() {
+    };
+
+    @Test
+    void withoutForClass() throws JsonProcessingException {
```

Comment:
Let's use `throws Exception`, going forward, just so merging to `master` is easier (exceptions are renamed plus no longer extend `IOException`).